Renable digest emails for soft deactivated users.

punchagan opened the pull request:
Digest emails were disabled for soft deactivated users, since UserMessage
objects are created for such users lazily when they return.

We now compute the message list for gathering hot conversations by looking at
all the messages sent to the streams where the user is subscribed, while they
were subscribed.

- Also includes a minor fix where context variables already computed are not updated again. 
- Also, a minor query optimization for computing the list of `home_view_recipients`

punchagan commented:
This is still a WIP.  Please don't review, yet.

punchagan commented:
This is now ready for review. I duplicated some of the code for computing missing messages for soft deactivated users. I'll work on refactoring the existing code, after an initial review of this PR.

timabbott commented on zerver/lib/digest.py:
```diff
@@ -230,11 +230,10 @@ def handle_digest_email(user_profile_id: int, cutoff: float,
+    home_view_recipients = Subscription.objects.filter(
+        user_profile=user_profile,
+        active=True,
+        in_home_view=True).values_list('recipient', flat=True)
```
I usually spell this as `recipient_id`, just to be clear it's integers and not objects that we want.

punchagan replied:
Fixed, thanks!

timabbott commented on zerver/tests/test_digest.py:
```diff
@@ -121,7 +121,7 @@ def test_multiple_stream_senders(self,
             handle_digest_email(othello.id, cutoff)
 
-        self.assertTrue(29 <= len(queries) <= 30)
+        self.assertTrue(24 <= len(queries) <= 25)
```
@showell you added this test in 51aa3138331e6596e713553280280b7c9d5478f7; do you remember why this was a range, not just a singular value?

punchagan replied:
I think the number of queries is different when we run this test alone vs all the tests in this class or the whole test-suite.

timabbott replied:
Well, I would have thought `flush_per_request_caches()` (combined with similar code for memcached and databases that we do for all tests) would have handled that issue.  But maybe not worth worrying about.

punchagan replied:
Yes, I wondered about that too, but didn't dig deeper. 

Running only `test_multiple_stream_senders` seems to need one query more than when running all the tests in `zerver.tests.test_digest.TestDigestEmailMessages`.

punchagan replied:
I think it's worth adding a comment explaining why there's a range, at least.

punchagan replied:
I did a quick diff and this seems like the additional query, when running the single test. 

```sql
SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."model" = 'userprofile' AND "django_content_type"."app_label" = 'zerver')
```

timabbott commented:
The preparatory refactors look good.  I guess one question that I'm thinking about is what duration we want to be constructing digests over (I don't remember how the code works).  It seems like for soft-deactivated users (who we know were offline for at least 3 weeks), we probably just want the digest to be constructed from the messages since the last digest was sent, and we probably want to just entirely exclude streams that the user is no longer subscribed to.  So possibly, we don't need to pull in the full `add_missing_messages` logic in order to make this work?

punchagan commented:
When trying to figure out who needs to be sent a digest email, the code looks at the users who were inactive for 5 (`DIGEST_CUTOFF`) days. We do look only for messages which were sent since the cutoff time (`now - DIGEST_CUTOFF`), and only the streams where the user is currently subscribed. But, there could be cases where the user's stream subscriptions were changed by other users or admins. In this case, without the full logic of `add_missing_messages` we might end up including messages which the user wouldn't receive in the digest.

timabbott commented:
Right, what I was thinking is that we could have much simpler logic that just excludes all streams this user has a RealmAuditLog subscribe/unsubscribe entry for during the period, and that might be a cleaner compromise than duplicating that logic (which, as a sidenote, is probably not super fast).

punchagan commented:
That does seem like a fair compromise. I'll update the PR!

punchagan commented:
I've pushed a change to simplify the logic for figuring out missing messages for soft deactivated users. I think this is ready for another review, now, @timabbott.

timabbott commented:
I merged the first couple commits; will review the rest later.

timabbott commented on zerver/lib/digest.py:
```diff
@@ -263,3 +264,45 @@ def handle_digest_email(user_profile_id: int, cutoff: float,
+        recipient__type_id__in=streams,
+        pub_date__gt=cutoff_date).select_related('recipient', 'sender', 'sending_client')
+
+    return messages
```
So, after reading this, I think that we could have a cleaner setup where this function return a set of `recipient_id` values, and just replaces the `home_view_recipients` argument used above (basically avoiding the `Message.objects.filter()` part of the code here being duplicated with the main code path).

punchagan replied:
I've refactored this. Thanks for the suggestion!

timabbott commented on zerver/tests/test_digest.py:
```diff
@@ -140,6 +127,61 @@ def test_multiple_stream_senders(self,
+        self.assertEqual(hot_convo['count'], 5 - 2)  # 5 messages, but 2 shown
+        teaser_messages = hot_convo['first_few_messages'][0]['senders']
+        self.assertIn('some content', teaser_messages[0]['content'][0]['plain'])
+        self.assertIn(teaser_messages[0]['sender'], expected_participants)
```
Maybe we just need a comment explaining this, but I can't quite figure out how this test ensures that (1) the streams we want to filter get excluded and (2) the other ones don't.  

Possibly with the refactor I suggest above, this could be replaced with something that's more of a unit test checking the recipient IDs list selected, which would definitely have that property.

punchagan replied:
Yes, that's indeed an insufficient test!

timabbott commented:
@punchagan I posted a couple comments; otherwise this looks great!  

I think once this bit of tooling is done, we might want to talk about making the `gather_hot_conversations` algorithm better.  E.g. number of emoji reactions would be a good factor for it to consider.  The current algorithm is pretty naive: Basically get some conversations with a lot of messages and a lot of senders.  I think we'd probably be better off having a scoring function for a conversation that has more of this sort of structure:
* +3 points per message in the thread (we don't want to over-weight this, because long effectively 1:1 support threads take over otherwise)
* +20 points per distinct sender
* +10 points per emoji reaction on messages in the thread
* +30 points if it's in one of your pinned streams.
* -50 points if the current user is mentioned (they already got an email)
* Etc.

(I gave very little thought to the actual numbers there; I just think conceptually we want this sort of structure for doing the selection that's based on scoring and picking the conversation(s) with the best score, rather than adding together lists from a couple mediocre algorithms).

punchagan commented:
Thanks for another review, @timabbott ! I will respond to and address your code specific comments when I have more time. 

Generally, the scoring scheme sounds like a great improvement over the current naive implementation. This is also similar to the ideas that @rishig had [here](https://chat.zulip.org/#narrow/stream/127-integrations/topic/ML.20Bot/near/536479) a while ago. 

Currently, we `gather_hot_conversations` per-user, but with a more complex scoring system, I wonder if we should have a first level of scoring that's user-independent, and then do a per-user scoring when sending the digest to an individual user.  Do you have thoughts on whether we should do that, and if yes, how to go about doing that?

timabbott commented:
Yes, I think we should eventually move to such an architecture.  I can imagine doing it directly or just doing it via caching of the computation of the score for a given stream/topic pair (or probably more likely, caching the mapping from stream -> (list of scored topics).

Any version of that sort of caching is likely going to benefit from the kind of "check the user's subscriptions didn't change" logic we're doing for soft-deactivated users, so that we don't need to worry about that sort of detail when doing the caching analysis.

punchagan commented:
@timabbott I've addressed the comments from your previous review. I refactored the code to figure out the streams we want to include for fetching the messages to include in the digest. Please take a look when you have some time.

rishig commented:
I'll copy out a summary of my thoughts from before. This was in the context of specifically coming up with a machine learning way to do this, but I think the basic idea or looking through a bunch of messages and getting some intuition from that holds whether or not we do it with ML. 

* Each email should include 3-6 messages from the last week's worth of messages on streams that the user is subscribed to. (We can talk about PMs separately.) The goal is to pick 3-6 messages that maximize the chance the user clicks through into the zulip realm.

To keep this sane (and shipable), the way I would do the V1 is:

* Restrict to picking messages that are the first message of their topics.
* For each such message, compute a user-independent "interestingness" score (more below)
* For each stream, compute the 3 messages with the highest interestingness score
* For each user, pick the 6 messages that have the highest interestingness score, among the streams they are subbed to.

For the interestingness step, I would
* Come up with a list of easy to compute features, like "length of message", "has images", "account age of person sending the message", "number of messages in thread", "number of messages in thread that have a lot of words", "number of emoji reactions", "has topic", "is default stream", "is private stream", etc.
* From czo and maybe other realms you're a part of, copy out the first messages from every topic, for every public stream you're a part of (we can discuss in PMs how to handle access to czo data)
* By hand, label a few hundred messages with good/bad, or with a sliding scale, of how interesting they are (this shouldn't take too long; I think I could do it myself in a few hours).

After doing the by-hand labelling, you may already have a pretty good intuition for what features matter. If not, we can do a basic regression to figure out which 2-3 features seem the most predictive and give scores based on that.

punchagan commented:
@timabbott This PR has a commit to re-enable digest emails for soft-deactivated users. Could you take a look at it? (It also has a minor comment commit)
I've removed some of the refactoring of the code to figure out what messages go into the digest email. This, along with improvements to the actual logic of picking messages for the digest can come in a different PR.

timabbott commented:
This looks good, merged, thanks @punchagan!

For the next PR: We don't have the real-time sync correct for updating the "notifications" setting tab after toggling this as the organization level.

timabbott commented on zerver/lib/digest.py:
```diff
@@ -166,36 +165,31 @@ def handle_digest_email(user_profile_id: int, cutoff: float,
-    home_view_recipients = Subscription.objects.filter(
+    home_view_streams = Subscription.objects.filter(
         user_profile=user_profile,
         active=True,
```
Let's add `recipient__type=Recipient.STREAM` to the filter here -- we don't want huddles etc. and `in_home_view=True` is sorta a weird way to express that.

punchagan replied:
I guess the code here is super old, and hence does it that way. I'll change it!

timabbott commented on zerver/tests/test_digest.py:
```diff
@@ -34,25 +34,16 @@ def test_multiple_stream_senders(self,
+        # One additional query is run to the content_type for the
+        # UserProfile model, when this test is run in isolation, as opposed to
+        # the whole class or test-suite.
+        self.assertTrue(6 <= len(queries) <= 7)
```
We can probably make this a fixed number by just doing whatever triggers that query the ling before the `with` block.

punchagan replied:
Agree! :+1:

timabbott commented on zerver/tests/test_digest.py:
```diff
@@ -71,6 +62,89 @@ def test_multiple_stream_senders(self,
+        self.simulate_stream_conversation('Scotland', senders)
+        self.simulate_stream_conversation('Denmark', senders)
+
+        one_sec_ago = timezone_now() - datetime.timedelta(seconds=1)
```
We should do an hour for this -- otherwise, the test will flake if the steps above took longer than a second, which they could on a slow/hosed machine.

punchagan commented:
@timabbott I've fixed this PR to address the comments from the last review. Thanks for thoughtful review and the suggested changes!

timabbott commented:
We may want to migrate https://github.com/zulip/zulip/pull/11761#issuecomment-474009923 to a new issue tracking work on improving the data for the emails.

punchagan commented:
> We may want to migrate [#11761 (comment)](https://github.com/zulip/zulip/pull/11761#issuecomment-474009923) to a new issue tracking work on improving the data for the emails.

Done in #12315
